docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+
+Putting these all together, we get the following partial lattice representing type promotion between Python scalars and numpy arrays:
+
+```{code-cell}
+:cellView: form
+:id: koA5VFHp7tjo
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+lattice = {
+  'i*': ['f*', 'u8', 'i8'], 'f*': ['c*', 'f16'], 'c*': ['c64'],
+  'u8': ['u16'], 'u16': ['u32'], 'u32': ['u64'],
+  'i8': ['i16'], 'i16': ['i32'], 'i32': ['i64'],
+  'f16': ['f32'], 'f32': ['f64'],
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [-0.5, 2], 'c*': [0, 3],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [0.5, 2], 'f32': [1.5, 2], 'f64': [2.5, 2],
+  'c64': [2, 3], 'c128': [3, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "WVjEMlmm9RS5"}
+
+Notice that this is not (yet) a true lattice: there are many pairs of nodes for which a join does not exist. However, we can think of this as a *partial* lattice, in which some pairs of nodes do not have a defined promotion behavior, and the defined portion of this partial lattice does correctly describe NumPy's array promotion behavior (leaving aside value-dependent semantics mentioned above).
+
+This sets up a nice framework by which we can think about filling-out these undefined promotion rules, by adding connections on this graph. But which connections to add?
+Broadly speaking, we want any additional connections to satisfy a few properties:
+
+1. Connections should maintain the commutativity and associativity of type promotion: that is, the graph should remain a (partial) lattice.
+
+2. Connections should never allow for a loss of data: for example, we should never promote `float` to `int`, as it would discard everything beyond the decimal point.
+
+3. Connections should never allow for a loss of magnitude. For example, the maximum possible `uint32` is twice as large as the maximum possible `int32`, so we should not implicitly promote `uint32` to `int32`.
+
+4. Wherever possible, connections should avoid loss of precision. For example, an `int64` value may have 64 bits of mantissa, so promoting `int64` to `float64` represents a possible loss of precision. However, because the maximum representable float64 is larger than the maximum representable int64, the *magnitude* can still be correctly represented.
+
+5. Wherever possible, connections should avoid unnecessary promotion to larger bit-widths. The purpose of this is to ensure that JAX's implicit promotions remain friendly to accelerator-based workflows, in which users often want to restrict types to 32-bit (or in some cases 16-bit) values.
```

Comment:
Go one step further: promotion will not produce higher precision types than those the user used as input.

Reply:
Done.